File: dotnet/src/Microsoft.Agents.Validation/Validators/ContentValidators/FunctionResultContentValidator.cs
using Microsoft.Agents.Abstractions.Models;

namespace Microsoft.Agents.Validation.Validators.ContentValidators;

/// <summary>
/// Validator for FunctionResultContent.
/// </summary>
public class FunctionResultContentValidator : ContentValidatorBase<FunctionResultContent>
{
    public override ValidationResult Validate(FunctionResultContent content, ValidationContext? context = null)
    {
        var errors = new List<ValidationError>();

        // REL-004: One FunctionResultContent per call-id (no duplicates)
        if (context != null && !string.IsNullOrEmpty(content.CallId))
        {
            if (context.FunctionResultExists(content.CallId))
            {
                errors.Add(CreateError(
                    ValidationErrorCode.REL_004,
                    $"Duplicate function result for call-id '{content.CallId}'. Each call-id can only have one function result.",
                    "CallId"));
            }
            else
            {
                context.RegisterFunctionResult(content);
            }

            // REL-002 & REL-006: FunctionResultContent call-id must match a FunctionCallContent call-id
            if (!context.FunctionCallExists(content.CallId))
            {
                errors.Add(CreateError(
                    ValidationErrorCode.REL_002,
                    $"Function result has call-id '{content.CallId}' but no matching function call was found",
                    "CallId"));
            }
            else
            {
                // REL-007: FunctionResultContent name must match FunctionCallContent name
                var functionCall = context.GetFunctionCall(content.CallId);
                if (functionCall != null &&
                    !string.IsNullOrEmpty(content.Name) &&
                    !string.IsNullOrEmpty(functionCall.Name) &&
                    content.Name != functionCall.Name)
                {
                    errors.Add(CreateError(
                        ValidationErrorCode.REL_007,
                        $"Function result name '{content.Name}' does not match function call name '{functionCall.Name}' for call-id '{content.CallId}'",
                        "Name"));
                }
            }
        }

        // CNT-005: FunctionResultContent result must be valid JSON
        var jsonError = ValidateJson(content.Result, "Result",
            ValidationErrorCode.CNT_005,
            "FunctionResultContent result must be valid JSON");
        if (jsonError != null)
            errors.Add(jsonError);

        return new ValidationResult(errors);
    }
}

File: typescript/packages/agents-protocol-validation/src/validators/content/FunctionResultContentValidator.ts
import type { FunctionResultContent } from '@microsoft/agents-protocol-abstractions';
import type { ValidationContext } from '../../core/ValidationContext';
import { ContentValidatorBase } from '../base/ContentValidatorBase';
import { ValidationResult } from '../../core/ValidationResult';
import { ValidationError } from '../../core/ValidationError';
import { ValidationErrorCode } from '../../core/ValidationErrorCode';

/**
 * Validator for FunctionResultContent.
 */
export class FunctionResultContentValidator extends ContentValidatorBase<FunctionResultContent> {
  public validate(content: FunctionResultContent, context?: ValidationContext): ValidationResult {
    const errors: ValidationError[] = [];

    // REL-002: call-id must be non-empty
    const callIdError = this.validateNotEmpty(
      content.callId,
      'callId',
      ValidationErrorCode.REL_002,
      'FunctionResultContent call-id must be non-empty'
    );
    if (callIdError) {
      errors.push(callIdError);
    }

    // REL-004: call-id must match a FunctionCall (if context available)
    if (context && content.callId) {
      if (!context.hasFunctionCall(content.callId)) {
        errors.push(
          new ValidationError(
            ValidationErrorCode.REL_004,
            `FunctionResultContent call-id does not match any FunctionCall: ${content.callId}`,
            'callId'
          )
        );
      }
    }

    // REL-006: call-id must be unique among results (if context available)
    if (context && content.callId) {
      if (!context.registerFunctionResult(content)) {
        errors.push(
          new ValidationError(
            ValidationErrorCode.REL_006,
            `Duplicate FunctionResultContent for call-id: ${content.callId}`,
            'callId'
          )
        );
      }
    }

    // CNT-005: result must be valid JSON (if present)
    if (content.result) {
      const resultError = this.validateJson(
        content.result,
        'result',
        ValidationErrorCode.CNT_005,
        'FunctionResultContent result must be valid JSON'
      );
      if (resultError) {
        errors.push(resultError);
      }
    }

    return new ValidationResult(errors);
  }
}
